src/main/java/com/simprints/simface/quality/MlKitFaceDetectionProcessor.kt

```diff
@@ -136,4 +146,258 @@ internal class MlKitFaceDetectionProcessor() : FaceDetectionProcessor {
+            noseContour.points.last()
+        } else {
+            return null
+        }
+
+        val lowerLipContour = face.getContour(FaceContour.LOWER_LIP_BOTTOM)
+        val mouthLeftLandmark = face.getLandmark(FaceLandmark.MOUTH_LEFT)
+        val mouthLeft = if (mouthLeftLandmark != null) {
+            mouthLeftLandmark.position
+        } else if (lowerLipContour != null && lowerLipContour.points.isNotEmpty()) {
+            lowerLipContour.points.last()
+        } else {
+            return null
+        }
+
+        val mouthRightLandmark = face.getLandmark(FaceLandmark.MOUTH_RIGHT)
+        val mouthRight = if (mouthRightLandmark != null) {
+            mouthRightLandmark.position
+        } else if (lowerLipContour != null && lowerLipContour.points.isNotEmpty()) {
+            lowerLipContour.points.first()
+        } else {
+            return null
+        }
+
+        return FacialLandmarks(
+            Point2D(leftEye.x, leftEye.y),
+            Point2D(rightEye.x, rightEye.y),
+            Point2D(nose.x, nose.y),
+            Point2D(mouthLeft.x, mouthLeft.y),
+            Point2D(mouthRight.x, mouthRight.y)
+        )
+    }
+
+
+    override fun warpAlignFace(face: FacialLandmarks, inputImage: Bitmap): Bitmap {
+
+        val ref = arrayOf(
+            floatArrayOf(38.2946f, 51.6963f),
+            floatArrayOf(73.5318f, 51.5014f),
+            floatArrayOf(56.0252f, 71.7366f),
+            floatArrayOf(41.5493f, 92.3655f),
+            floatArrayOf(70.7299f, 92.2041f)
+        )
+
+        val landmarks = landmarkSetToArray(face)
+        val tfm = computeTFM(landmarks, ref)
+        val warped = warpAffine(inputImage, tfm)
+
+        return warped
+    }
+
+    fun warpAffine(
+        inputBitmap: Bitmap,
+        tfm: SimpleMatrix,   // 2x3 matrix from EJML
+        cropWidth: Int = 112,
+        cropHeight: Int = 112
+    ): Bitmap {
+
+        val outputBitmap = Bitmap.createBitmap(cropWidth, cropHeight, Bitmap.Config.ARGB_8888)
+
+        // Extract matrix values from tfm (2x3)
+        val values = FloatArray(9) { 0f }
+        values[0] = tfm[0, 0].toFloat() // scale X
+        values[1] = tfm[0, 1].toFloat() // skew X
+        values[2] = tfm[0, 2].toFloat() // translate X
+        values[3] = tfm[1, 0].toFloat() // skew Y
+        values[4] = tfm[1, 1].toFloat() // scale Y
+        values[5] = tfm[1, 2].toFloat() // translate Y
+        values[6] = 0f // perspective X
+        values[7] = 0f // perspective Y
+        values[8] = 1f // perspective scale
+
+        // Create Android Matrix with the transformation
+        val androidMatrix = Matrix().apply {
+            setValues(values)
+        }
+
+        // Apply transformation to canvas
+        val canvas = Canvas(outputBitmap)
+        canvas.drawBitmap(inputBitmap, androidMatrix, null)
+
+        return outputBitmap
+    }
+
+    fun findNonreflectiveSimilarity(fac: Array<FloatArray>, ref: Array<FloatArray>): SimpleMatrix {
+        val K = 2
```

Comment:
Please use full descriptive names or add a comment that explains what is going on. Not every reader of the code remembers all of the details of the underlying math.